client/http/client.go

```diff
@@ -147,16 +147,18 @@ func (ci *clientInner) requestWithRetry(
 	headerOpts ...HeaderOption,
 ) error {
 	var (
+		statusCode             int
 		err                    error
 		addr                   string
 		pdAddrs, leaderAddrIdx = ci.getPDAddrs()
 	)
 	// Try to send the request to the PD leader first.
 	if leaderAddrIdx != -1 {
 		addr = pdAddrs[leaderAddrIdx]
-		err = ci.doRequest(ctx, addr, reqInfo, headerOpts...)
-		if err == nil {
-			return nil
+		statusCode, err = ci.doRequest(ctx, addr, reqInfo, headerOpts...)
+		// No need to retry if there is no error or the status code is 404 returned by the PD leader.
+		if err == nil || statusCode == http.StatusNotFound {
```

Comment:
Also for StatusInternalServerError and StatusBadRequest?

Reply:
IMO, only the latter one is no need.

Reply:
BTW, we also have API which will return the 403 error.

Reply:
Or we discuss what situations need retries?

Reply:
We can first exclude some obvious scenarios that do not require retrying, and then supplement more later.